src/wrapper.c

```diff
@@ -1,551 +1,563 @@
-		puts("FAILED (crypt_gensalt)\n");
-		return 1;
-	}
-
-	setting2 = crypt_gensalt_ra(setting1, 12, data, size);
-	if (strcmp(setting1, setting2)) {
-		puts("FAILED (crypt_gensalt_ra/1)\n");
-		return 1;
-	}
-
-	(*(char *)data)++;
-	setting1 = crypt_gensalt_ra(setting2, 12, data, size);
-	if (!strcmp(setting1, setting2)) {
-		puts("FAILED (crypt_gensalt_ra/2)\n");
-		return 1;
-	}
-
-	free(setting1);
-	free(setting2);
-	free(data);
-
-#if defined(_SC_CLK_TCK) || !defined(CLK_TCK)
-	clk_tck = sysconf(_SC_CLK_TCK);
-#else
-	clk_tck = CLK_TCK;
-#endif
-
-	running = 1;
-	signal(SIGALRM, handle_timer);
-
-	memset(&it, 0, sizeof(it));
-	it.it_value.tv_sec = 5;
-	setitimer(ITIMER_REAL, &it, NULL);
-
-	start_real = times(&buf);
-	start_virtual = buf.tms_utime + buf.tms_stime;
-
-	count = (char *)run((char *)0) - (char *)0;
-
-	end_real = times(&buf);
-	end_virtual = buf.tms_utime + buf.tms_stime;
-	if (end_virtual == start_virtual) end_virtual++;
-
-	printf("%.1f c/s real, %.1f c/s virtual\n",
-		(float)count * clk_tck / (end_real - start_real),
-		(float)count * clk_tck / (end_virtual - start_virtual));
-
-#ifdef TEST_THREADS
-	running = 1;
-	it.it_value.tv_sec = 60;
-	setitimer(ITIMER_REAL, &it, NULL);
-	start_real = times(&buf);
-
-	for (i = 0; i < TEST_THREADS; i++)
-	if (pthread_create(&t[i], NULL, run, i + (char *)0)) {
-		perror("pthread_create");
-		return 1;
-	}
-
-	for (i = 0; i < TEST_THREADS; i++) {
-		if (pthread_join(t[i], &t_retval)) {
-			perror("pthread_join");
-			continue;
-		}
-		if (!t_retval) continue;
-		count = (char *)t_retval - (char *)0;
-		end_real = times(&buf);
-		printf("%d: %.1f c/s real\n", i,
-			(float)count * clk_tck / (end_real - start_real));
-	}
-#endif
-
-	return 0;
-}
-#endif
+/*
+ * Written by Solar Designer <solar at openwall.com> in 2000-2014.
+ * No copyright is claimed, and the software is hereby placed in the public
+ * domain.  In case this attempt to disclaim copyright and place the software
+ * in the public domain is deemed null and void, then the software is
+ * Copyright (c) 2000-2014 Solar Designer and it is hereby released to the
+ * general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * There's ABSOLUTELY NO WARRANTY, express or implied.
+ *
+ * See crypt_blowfish.c for more information.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <errno.h>
+#ifndef __set_errno
+#define __set_errno(val) errno = (val)
+#endif
+
+#ifdef TEST
+#include <stdio.h>
+#include <unistd.h>
+#include <signal.h>
+#include <time.h>
+#include <sys/time.h>
+#include <sys/times.h>
+#ifdef TEST_THREADS
+#include <pthread.h>
+#endif
+#endif
+
+#define CRYPT_OUTPUT_SIZE		(7 + 22 + 31 + 1)
+#define CRYPT_GENSALT_OUTPUT_SIZE	(7 + 22 + 1)
+
+#if defined(__GLIBC__) && defined(_LIBC)
+#define __SKIP_GNU
+#endif
+
+#ifdef _WIN32 | _WIN64
+#include "../include/bcrypt/ow-crypt.h"

```

Comment:
Same to this, why should we use relative paths? Is this required by Visual Studio or something?

Reply:
I think its nicer, but can switch it over to <> if you want no worries.